### PR TITLE
Allow for resolver extension

### PR DESCRIPTION
**Pull request by agjohnson**
Allow for extension of resolver methods by moving everything into an extendable class.

While my dev environment was hosed and was waiting for Python and all my tooling to compile/install, I wrote up a nicer implementation of lazy classes for overriding with a setting. The main benefit here is that we only need two classes to allow for override, and they can exist in the same source file. We can also automate the logic around where to pull the override setting from. There were several issues that I hit while trying to expose the class methods as module functions, this solved those issues in a much cleaner way. Use of `CLASS_OVERRIDES` is for the future, so we can have an explicit list of the Classes overridden instead.

The diff on this looks awful, so I'll enumerate what was actually changed, besides moving everything into a class:
* `_get_canonical_project` was broken out into a method
* `_get_project_subdomain` was broken out into a method
* Everything was made to reference a class instance with self
* `_use_subdomain` logic was broken out and updated on the calls i was changing

One piece of logic changed on the resolver: if `USE_SUBDOMAIN` is set without a `PUBLIC_DOMAIN` setting, don't treat `PRODUCTION_DOMAIN` as the subdomain hosting domain. Instead, default to serving from `/docs/...`. This is reflected in the test changes.

### CODE REVIEW

**Comment by agjohnson** [comment]
WIP here too, have some changes inbound

**Comment by agjohnson** [comment]
Okay, tests fixed, ready for review. Note updated ticket description.

**Comment by ericholscher** [comment]
Seems like if `USE_SUBDOMAIN` is set, we'd not want to serve from `/docs/` URL.. what is the reasoning behind that change?

**Comment by ericholscher** [comment]
Ah, I see in the tests -- we were serving from the subdomain, but with the `/docs/` URL -- in general I think we should bias against any use of the `/docs/` URL, only using it when necessary in dev -- and we should be able to get away from that with a remote wildcard domain that points at 127.0.0.1.

File: readthedocs/core/utils/extend.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+"""Patterns for extending Read the Docs"""
+
+from django.conf import settings
+from django.utils.module_loading import import_by_path
+from django.utils.functional import LazyObject
+
+
+class SettingsOverrideObject(LazyObject):
```
Comment (ericholscher) [inline readthedocs/core/utils/extend.py]: Would be good to see a test of this class -- mainly just to wrap my head around the usage. I guess most of the logic is enabled by Django's `LazyObject`, but might be good to at least test our integration.

Comment (agjohnson) [inline readthedocs/core/utils/extend.py]: Yeah, easy enough

**Comment by ericholscher** [comment]
Looks good overall -- mainly just have nitpicks about resolution, but the current implementation is good for now.

**Comment by agjohnson** [comment]
Yeah, reason behind the change was that we don't want to serve subdomain docs from `PRODUCTION_DOMAIN` under any circumstances. I'm not sure it makes sense to drop `USE_SUBDOMAIN` for `if PUBLIC_DOMAIN is not None:` sort of logic, as that's not horribly clear either. It seems the most correct thing is to have `USE_SUBDOMAIN` be explicitly dependant on `PUBLIC_DOMAIN` -- possibly to the point we throw an exception?

With changes to how we view subdomain hosting, I think it makes sense to require both. It may push independent installs to use `/docs/` but I don't think this does much else to enable serving from `/docs/`

**Pull request merged**